lib/TBDGen/TBDGen.cpp

```diff
@@ -496,10 +496,15 @@ void TBDGenVisitor::addConformances(DeclContext *DC) {
                 reqtAccessor->getAccessorKind());
             addSymbolIfNecessary(reqtAccessor, witnessAccessor);
           });
-        } else if (isa<EnumElementDecl>(witnessDecl)) {
-          addSymbolIfNecessary(valueReq, witnessDecl);
         }
       }
+
+      // Under -enable-testing, the enum case constructor symbols must be
+      // exported.
+      if (isa<EnumElementDecl>(witnessDecl) &&
+          SwiftModule->isTestingEnabled()) {
+        addSymbol(SILDeclRef(witnessDecl, SILDeclRef::Kind::EnumElement));
```

Comment:
I don't think this is the best fix. The case constructors should not be public with -enable-testing. I think they need to be emitted with shared linkage, not hidden, in the case that the enum is `internal` at the AST level. Can you take a look at SILDeclRef::getLinkage() and see if you can fix it there?

Reply:
Thank you for the suggestion! Adding a check to `getLinkage` has no effect unfortunately:

```c++
if (isEnumElement() && d->getFormalAccess() == AccessLevel::Internal) {
  return SILLinkage::Shared;
}
```

Reply:
Sorry, I forgot I made the protocol and enum `public`. When I make them `internal`, I get a crash instead in `addConformances` because its calling `mangleWitnessThunk` which expects the requirement to be a function or constructor. Removing the call to `addSymbolIfNecessary` fixes the problem but then you get the same missing symbol error for `$s4test1EOAA1PA2aDP4bar1xvgZTW` which is the witness thunk.

Reply:
Enum elements need to be handled the same as on-demand accessors. See this block:

```
  // Forced-static-dispatch functions are created on-demand and have
  // at best shared linkage.
  if (auto fn = dyn_cast<FuncDecl>(d)) {
    if (fn->hasForcedStaticDispatch()) {
      limit = Limit::OnDemand;
    }
  }
```

You just need to add another check, like

```
if (isEnumElement())
   limit = Limit::OnDemand;
```

Reply:
Hmm,  I added this exact check before (commit https://github.com/apple/swift/pull/28916/commits/7d3cf178f125c07dd3908cfb86cb884d94c0cea2) but I removed it for some reason 🤷

Reply:
Ah, nice catch! Yeah, let's bring this back, and cherry-pick the fix back to 5.3 too.

Reply:
Okay, I have added those checks. But, it seems like the combination of the checks is preventing `TBDGenVisitor::addConformances` from adding the symbol for the witness of the property requirement (which is an enum case without payload). I can't use `addSymbolIfNecessary` anymore if the requirement is a property and adding a symbol by passing the mangled name doesn't seem to work as well.